cpp/src/arrow/dataset/discovery.cc

```diff
@@ -244,7 +244,7 @@ Result<std::vector<std::shared_ptr<Schema>>> FileSystemDatasetFactory::InspectSc
     if (ARROW_PREDICT_FALSE(!result.ok())) {
       return result.status().WithMessage(
           "Error creating dataset. Could not read schema from '", info.path(),
-          "': ", result.status().message(), ". Is this a '", format_->type_name(),
+          "': ", result.status().message(), " Is this a '", format_->type_name(),
```

Comment:
Minor bike-shedding nit.  Should we surround the inner exception with `()`?

```
pyarrow.lib.ArrowInvalid: Error creating dataset. Could not read schema from 'data/flights-200k.arrow': (Could not open Parquet input source 'data/flights-200k.arrow': Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.) Is this a 'parquet' file?
```

Reply:
I agree, I think `()` would help with emphasizing an inner exception.
Additionally, I see file type repeated, may I modify the last line of the inner exception, like so?

`pyarrow.lib.ArrowInvalid: Error creating dataset. Could not read schema from 'data/flights-200k.arrow': (Could not open Parquet input source 'data/flights-200k.arrow': Parquet magic bytes not found in footer. This file might be corrupted.) Is this a 'parquet' file?`

P.S. I might also need to make respective change in the Go component.

Reply:
Hmm, the inner exception can occur in other contexts too though.  For example, when reading a single parquet file without datasets.  In that case the idea that it might not be a parquet file could be helpful.  I think the double-sentence might be preferred still.

Reply:
How about changing order instead of using `()`?

```diff
diff --git a/cpp/src/arrow/dataset/discovery.cc b/cpp/src/arrow/dataset/discovery.cc
index a38ec00bb9..2aca85e6ad 100644
--- a/cpp/src/arrow/dataset/discovery.cc
+++ b/cpp/src/arrow/dataset/discovery.cc
@@ -244,8 +244,7 @@ Result<std::vector<std::shared_ptr<Schema>>> FileSystemDatasetFactory::InspectSc
     if (ARROW_PREDICT_FALSE(!result.ok())) {
       return result.status().WithMessage(
           "Error creating dataset. Could not read schema from '", info.path(),
-          "': ", result.status().message(), ". Is this a '", format_->type_name(),
-          "' file?");
+          "'. Is this a '", format_->type_name(), "' file?: ", result.status().message());
     }
     schemas.push_back(result.MoveValueUnsafe());
   }
```

Reply:
@kou From a developer's perspective the inner exception msg showing up in the middle makes sense for me, holistically.

That said, I do not have a deep understanding of the project at the moment :sweat_smile:

@westonpace Thoughts?

Reply:
@westonpace Ah okay, sounds good :+1:  I'll incorporate both the suggestions and push shortly.

Reply:
Change pushed

Reply:
Checking tests

Reply:
Please excuse the delay on this from my end, I will update this PR in a day or two, thank you.